### ISSUE TITLE
Reproduce Shadrikov's scheme 1: pretrain + classification (finetuning on FIW)

### ISSUE THREAD

**Issue by vitalwarley** [vitalwarley/research#20]
> First, we tried to add a simple **classification** layer and **finetune** the whole model on the train set with stochastic gradient descent with base learning rate of 0.0001, momentum 0.9, linear warmup for first 200 batches of size 64, linear cooldown for last 400 batches, multiplying learning rate by 0.75 on epochs 8, 14, 25, 35, 40 and gradient clipping 1.5 for 50 epochs. Random color jitter and random lightning with parameter 0.15 were used for the data augmentation. No random cropping or similar technique was used to not confuse the model that was trained on similar aligned images. 

The emphasis above is my own.

For this task, a sample is `(image, familiy_label)`. The `KinshipDataModule` already defines `train_ds` as the `FamiliesDataset`, 

```python
        if stage == "fit":
            self.train_ds = FamiliesDataset(
                self.data_dir / "train-faces-det", transform=self.train_transform
            )
```

which returns: `img, family_idx, person_idx`.

**Comment by vitalwarley** [comment]
There are some discrepancies between the paper and the `fitw2020/train.py`

```python
...
    batch_size = 48  # paper says 64
...
    lr = 1e-4  # paper says correctly
...
    num_epoch = 20  # paper says 50 epochs    momentum = 0.9

...
    wd = 1e-4  # paper doesn't mention it
    clip_gradient = 1.0  # paper says 1.5
    lr_steps = [8, 14, 25, 35, 40, 50, 60]  # paper doesn't include 50, and 60
...
    if warmup > 0:
        start_lr = 1e-10  # paper says 1e-4
    else:
        start_lr = lr
...
    end_lr = 1e-10  # paper doesn't mention it
```

Also, regarding the loss function

```python
    all_losses = [
                  ('softmax', gluon.loss.SoftmaxCrossEntropyLoss()),
                  # ('arc', gluonfr.loss.ArcLoss(num_families, m=0.7, s=32, easy_margin=False)), 
                  # ('center', gluonfr.loss.CenterLoss(num_families, 512, 1e-1))
                  ]

```

the paper isn't specific about what was used.

**Comment by vitalwarley** [comment]
My first experiment with a classification layer didn't go well

```python
  (fc): Linear(in_features=32768, out_features=512, bias=True)
  (bn): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  **(classification): Linear(in_features=512, out_features=763, bias=True)**
  **(loss): CrossEntropyLoss()**
````

where the experiment flags where

```
flags:
  accumulate_grad_batches: 1
  batch-size: 64
  cooldown: 400
  data-dir: /home/warley/dev/datasets/fiw
  devices: 1
  embedding-dim: 512
  end-lr: 1.0e-10
  gradient_clip_val: 1.5
  loss: ce
  lr: 1.0e-10
  lr-factor: 0.75
  lr-steps: 8, 14, 25, 35, 40, 50, 60
  max_epochs: 50
  model: resnet101
  momentum: 0.9
  normalize: no
  nprocs: 1
  precision: 16
  scheduler: multistep
  task: finetune
  warmup: 200
  weight-decay: 0.0001
scalars:
  lightning_logs/version_0#accuracy/train: 0.000000 (step 12399)
  lightning_logs/version_0#accuracy/val: 0.500200 (step 12399)
  lightning_logs/version_0#auc/val: 0.539312 (step 12399)
  lightning_logs/version_0#epoch: 49.000000 (step 12399)
  lightning_logs/version_0#hp_metric: -1.000000 (step 0)
  lightning_logs/version_0#loss/train: 6.647321 (step 12399)
  lightning_logs/version_0#lr: 0.000000 (step 12399)
```

What I noticed

- The loss didn't change -- nor decreased or increased consistently. 
- The LR scheduler seems to be not working -- the LR remained constant throughout the training.
  - Maybe `optimizer_step`  needs to be checked.
- Accuracy, therefore, also didn't change -- after the above items, that is to be expected.
  - However, the accuracy logic for validation needs to be improved. I am using a threshold of 0.5 for similarity, but as we know (#18), the right threshold is 
  - For training accuracy, hm, maybe something is missing? I will check the LR scheduler and try again.

### TODO

- [ ] Check LR scheduling logic and `optimizer_step` 
- [ ] Update validation accuracy: get `log_results`  from `val_shadrikov.py`

**Comment by vitalwarley** [comment]
I found a problem with a parameter. `lr_steps` was being passed in a wrong format: `['8, 14, 25, 35, 40, 50, 60']`

**Comment by vitalwarley** [comment]
![2022-07-19-045838_1457x546_scrot](https://user-images.githubusercontent.com/6365065/179698225-e138ad93-933f-400a-af0c-25bfc90b4a3c.png)
![2022-07-19-050001_1060x719_scrot](https://user-images.githubusercontent.com/6365065/179698490-9bb765d0-fed5-404b-940f-914c6489bd59.png)

LR OK, but the results didn't change much.

**Comment by vitalwarley** [comment]
I tried insightface's pretrained backbone and got better results, but not what I expected... yet.

![2022-07-19-111026_1096x714_scrot](https://user-images.githubusercontent.com/6365065/179771120-36512b88-5bab-42f6-b709-abb9de3d46a8.png)

**Comment by vitalwarley** [comment]
- [ ] Run `fitw2020/train.py` to confirm convergence of his scheme.

**Comment by vitalwarley** [comment]
> I don't do a softmax activation. Rather, I used the logits as is.

It is done by [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html).

> The input is expected to contain raw, unnormalized scores for each class.
> ...
> The target that this criterion expects should contain either:
> - Class indices in the range [0, C) where C is the number of classes; ...
> Note that this case is equivalent to the combination of [LogSoftmax](https://pytorch.org/docs/stable/generated/torch.nn.LogSoftmax.html#torch.nn.LogSoftmax) and [NLLLoss](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html#torch.nn.NLLLoss).

**Comment by vitalwarley** [comment]
> * [x]  Run `fitw2020/train.py` to confirm convergence of his scheme.

![2022-07-20-100909_606x600_scrot](https://user-images.githubusercontent.com/6365065/179996096-09e25089-60db-4526-8ced-7d24ac2b02a9.png)

Accuracy goes to 0.64 and loss to 2.373e-5 at epoch 40. Training was supposed to last 50 epochs, but I made a mistake. Also, I couldn't compute AUC with the save params. I will train again later as well as try to compute AUC. 

This result indicates something wrong with my model's or the dataset. I will check them, but the model is the most probable.

**Comment by vitalwarley** [comment]
> Accuracy goes to 0.64 and loss to 2.373e-5 at epoch 40. Training was supposed to last 50 epochs, but I made a mistake. Also, I couldn't compute AUC with the save params. I will train again later as well as try to compute AUC.

Well, here is the results:

```
[Epoch 049] softmax: 0.827| accuracy: 0.953672102136266 | time: 264.5
```

**Comment by vitalwarley** [comment]
> I will train again with original batch size on RIG3 today.

![2022-07-23-081738_1549x368_scrot](https://user-images.githubusercontent.com/6365065/180602795-7f2d28b8-009d-4901-9b0d-a3f4ea47da0b.png)
![2022-07-23-081712_1528x382_scrot](https://user-images.githubusercontent.com/6365065/180602797-5aff3860-441e-48d5-bbc8-c227f2433bd2.png)

- Original batch size is in blue. Accuracy was 0.538 and loss was 3.78.
- The red curve was using my datamodule, which means that the problem we are seeing isn't the result of the data. One more evidence that points to the model. 
- Also, the problem isn't how accuracy is computed, because I used torchmetrics in shadrikov script using logits (like shadrikov did) or preds from argmax (like I did in my model) and the results were the same.
- Lastly, validation results from mxnet (see above) are somewhat equivalent to my model's (after finetuning). I got 0.69, 0.755, and 0.07 for valitation accuracy, AUC and verification threshold, respectively.

All above indicates some problem in the training step. It seems like it is in the classification layer, mainly because of the validation results, which are computed using the embeddings.

**Comment by vitalwarley** [comment]
### Torch

![2022-07-25-095645_1459x696_scrot](https://user-images.githubusercontent.com/6365065/180782823-908da67a-e316-4504-9df6-78a33c27fca5.png)

#### Flags and results

```
flags:
  accumulate_grad_batches: 1
  batch-size: 16
  cooldown: 400
  data-dir: /home/warley/dev/datasets/fiw
  devices: '1'
  embedding-dim: 512
  end-lr: 1.0e-10
  gpus: 1
  gradient_clip_val: 5
  insightface: yes
  loss: ce
  lr: 0.0001
  lr-factor: 0.75
  lr-steps: 8 14 25 35 40 50 60
  max_epochs: 50
  model: resnet101
  momentum: 0.9
  normalize: no
  num-classes: 570
  precision: 16
  scheduler: multistep
  start-lr: 1.0e-10
  task: finetune
  warmup: 200
  weight-decay: 0.0001
  weights: /home/warley/dev/research/weights/arcface_torch.pth
scalars:
  lightning_logs/version_0#epoch: 49.000000 (step 49449)
  lightning_logs/version_0#hp_metric: -1.000000 (step 0)
  lightning_logs/version_0#lr: 0.000015 (step 49449)
  lightning_logs/version_0#train/accuracy: 0.878334 (step 49449)
  lightning_logs/version_0#train/loss: 2.396770 (step 49449)
  lightning_logs/version_0#val/accuracy: 0.679900 (step 49449)
  lightning_logs/version_0#val/auc: 0.747001 (step 49449)
  lightning_logs/version_0#val/threshold: 0.060000 (step 49449)
```
### MXNet

![2022-07-25-095633_1062x717_scrot](https://user-images.githubusercontent.com/6365065/180782833-9090adf0-774f-46a6-9d4d-2189a0d95eec.png)

#### Flags and results

```
flags:
  batch_size: 16
  clip_gradient: 1.0
  cooldown: 400
  end_lr: 1.0e-10
  gpu: 1
  jitter_param: 0.15
  lighting_param: 0.15
  lr: 0.0001
  lr_factor: 0.75
  lr_steps: 8 14 25 35 40 50 60
  model_name: arcface_r100_v1
  momentum: 0.9
  net_name: arcface_families_ft
  normalize: no
  num_epoch: 50
  num_workers: 12
  warmup: 200
  wd: 0.0001
scalars:
  logs/1658604570#lr: 0.000000 (step 49)
  logs/1658604570#train/accuracy: 0.937998 (step 49)
  logs/1658604570#train/loss: 0.957331 (step 49)
  logs/1658604570#val/accuracy: 0.708900 (step 49)
  logs/1658604570#val/roc/auc: 0.775784 (step 49)
  logs/1658604570#val/threshold: 0.110000 (step 49)
```

**Comment by vitalwarley** [comment]
From the above results, we can conclude:
- `batch-size` was less than what Shadrikov said it to be.
- `gradient_clip_val` needed to be higher than what Shadrikov said -- I used 5 because that was the value I pretrained the model on MS1M. Also, he said 1.5 on paper but wrote 1.0 on code.
- My results were close to Shadrikov's. I think my model would improve on train metrics (classification), but maybe not on validation metrics (verification).

Training with torch or mxnet requires different hparams, so I don't think I can reproduce 100% of his work.
